File: backend/src/utils/pagination.js
/**
 * Create pagination object for MongoDB queries
 * @param {Object} query - Express request query object
 * @param {Object} options - Additional options
 * @param {number} options.defaultPage - Default page number (default: 1)
 * @param {number} options.defaultLimit - Default items per page (default: 10)
 * @returns {Object} Pagination configuration object
 */
const getPagination = (query, options = {}) => {
  const page = Math.max(1, parseInt(query.page) || options.defaultPage || 1)
  const limit = Math.max(1, parseInt(query.limit) || options.defaultLimit || 10)
  const skip = (page - 1) * limit

  return {
    page,
    limit,
    skip
  }
}

/**
 * Create pagination metadata
 * @param {number} total - Total number of items
 * @param {number} page - Current page
 * @param {number} limit - Items per page
 * @returns {Object} Pagination metadata
 */
const getPaginationMetadata = (total, page, limit) => {
  return {
    total,
    page,
    limit,
    totalPages: Math.ceil(total / limit),
    hasNextPage: page < Math.ceil(total / limit),
    hasPrevPage: page > 1
  }
}

/**
 * Create paginated response
 * @param {Array} data - Array of items for current page
 * @param {number} total - Total number of items
 * @param {Object} pagination - Pagination configuration from getPagination
 * @returns {Object} Formatted response with data and pagination metadata
 */
const paginatedResponse = (data, total, pagination) => {
  return {
    data,
    pagination: getPaginationMetadata(total, pagination.page, pagination.limit)
  }
}

module.exports = {
  getPagination,
  getPaginationMetadata,
  paginatedResponse
}

File: backend.NET/Utils/Pagination.cs
namespace api.Utils
{
    public class Pagination
    {
        public int Page { get; set; }
        public int Limit { get; set; }
        public int Skip => (Page - 1) * Limit;

        public Pagination(int? page, int? limit, int defaultPage = 1, int defaultLimit = 10)
        {
            Page = page.HasValue && page.Value > 0 ? page.Value : defaultPage;
            Limit = limit.HasValue && limit.Value > 0 ? limit.Value : defaultLimit;
        }
    }

    public class PaginationMetadata
    {
        public int Total { get; set; }
        public int Page { get; set; }
        public int Limit { get; set; }
        public int TotalPages { get; set; }
        public bool HasNextPage { get; set; }
        public bool HasPrevPage { get; set; }

        public PaginationMetadata(int total, int page, int limit)
        {
            Total = total;
            Page = page;
            Limit = limit;
            TotalPages = (int)Math.Ceiling((double)total / limit);
            HasNextPage = page < TotalPages;
            HasPrevPage = page > 1;
        }
    }

    public class PaginatedResponse<T>
    {
        public IEnumerable<T> Data { get; set; }
        public PaginationMetadata Pagination { get; set; }

        public PaginatedResponse(IEnumerable<T> data, int total, Pagination pagination)
        {
            Data = data;
            Pagination = new PaginationMetadata(total, pagination.Page, pagination.Limit);
        }
    }
}
